zorch/lnp/exact.py

```
@@ -0,0 +1,229 @@
+        # reads the digits back as the number they encode. Public, so it is
+        # the half that carries the automorphism, exactly as `T`'s first
+        # argument does everywhere in this package.
+        radix = np.zeros((1, ring.d), dtype=np.int64)
+        radix[0, :digits] = 1 << np.arange(digits)
+        self._radix = ring.galois(ring.from_signed_stack(radix), 2 * ring.d - 1)[0]
+        # `1⃗` over one ring element: `Σ_t X^t`, the all-ones coefficient
+        # vector Lemma 5.2 subtracts.
+        self._ones = ring.from_signed_stack(np.ones((1, ring.d), dtype=np.int64))[0]
+
+    def decompose(self, witness: np.ndarray) -> np.ndarray:
+        """The digits `x⃗` of `β² − ‖(s1, m)‖²`, as a signed-integer stack the
```

Ruff (0.16.2)

[warning] Docstring contains ambiguous `−` (MINUS SIGN). Did you mean `-` (HYPHEN-MINUS)?

(RUF002)